misk-hibernate/src/main/kotlin/misk/hibernate/BoxedStringType.kt

```diff
@@ -0,0 +1,105 @@
+import kotlin.reflect.full.memberProperties
+import kotlin.reflect.full.primaryConstructor
+
+/**
+ * Binds any data class that wraps a String to a varchar in MySQL. This is most useful to enforce
+ * additional type-safety on identifier columns.
+ *
+ * To use, create a data class with a single non-null `String` property. The name of this property
+ * is not significant.
+ *
+ * ```
+ * data class MovieToken(val token: String)
+ * ```
+ *
+ * To use with comparison operators (ie. WHERE clauses with `<` or `>`) you must also implement
+ * `Comparable`:
+ *
+ * ```
+ * data class MovieToken(val token: String) : Comparable<MovieToken> {
+ *   override fun compareTo(other: MovieToken) = token.compareTo(other.token)
+ * }
+ * ```
+ */
+internal class BoxedStringType<T : Any> private constructor(
+  private val property: KProperty1<T, String>,
+  private val constructor: KFunction<T>
+) : UserType {
+  override fun hashCode(x: Any) = x.hashCode()
+
+  override fun deepCopy(value: Any) = value
+
+  override fun replace(original: Any, target: Any, owner: Any?) = original
+
+  override fun equals(x: Any, y: Any) = x == y
+
+  override fun returnedClass() = constructor.returnType.javaClass
+
+  override fun assemble(cached: Serializable, owner: Any?) = constructor.call(cached)
+
+  @Suppress("UNCHECKED_CAST") // Hibernate promises to call us only with the types we support.
+  override fun disassemble(value: Any?) = property.get(value as T)
+
+  override fun nullSafeSet(
+    st: PreparedStatement,
+    value: Any?,
+    index: Int,
+    session: SharedSessionContractImplementor?
+  ) {
+    if (value == null) {
+      st.setNull(index, Types.VARCHAR)
+    } else {
+      st.setString(index, disassemble(value))
+    }
+  }
+
+  override fun nullSafeGet(
+    rs: ResultSet,
+    names: Array<out String>,
+    session: SharedSessionContractImplementor?,
+    owner: Any?
+  ): Any? {
+    val result = rs.getString(names[0])
+    return if (result != null) assemble(result as String, owner) else null
+  }
+
+  override fun isMutable() = false
+
+  override fun sqlTypes() = intArrayOf(Types.VARCHAR)
+
+  companion object {
+    /**
+     * Returns a boxed string type if `propertyType` is a data class with a single string property.
+     * Otherwise this returns null.
+     */
+    fun <T : Any> create(propertyType: KClass<T>): BoxedStringType<*>? {
+      if (!propertyType.isData) return null
+
+      val memberPropertiesList = propertyType.memberProperties.toList()
+      if (memberPropertiesList.size != 1) return null
+
+      val onlyProperty: KProperty1<T, *> = memberPropertiesList[0]
+      if (onlyProperty.returnType.classifier != String::class) return null
+      @Suppress("UNCHECKED_CAST") // Guarded by a runtime check above.
+      val stringProperty = onlyProperty as KProperty1<T, String>
+
+      val constructor: KFunction<T>? = propertyType.primaryConstructor
+      if (constructor == null) return null // No primary constructor.
```

Comment:
nit: could be kotlin-y and use a `return when`

Reply:
Will try in follow up.